CICE-Consortium/CICE
# Update testing, in particular for diagnostics and decompositions

For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>


## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Update testing, in particular for diagnostics and decompositions
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Full test suite run on cheyenne with 3 compilers.  Tests are bit-for-bit except boxslotcyl which had a change to the namelist settings (tr_lvl=tr_pond_lvl=.false.) and the calchk unit test which was set to 1000 years of testing by default, but should have been 100,000 years.  Several new tests were added.
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit (except as noted above)
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
- Does this PR add any new test cases?
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] No, does the documentation need to be updated at a later time?
- [X] Please provide any additional information or relevant details below:

List of changes are below.  Of particular note
 - a couple bugs were fixed for sections of the code that we were generally not using/testing but are now
     - error in size of allocation in 1d rake
     - out of bounds issue for fcontopn_d and fsurfn_d when ncat > 6
 - the dbug namelist was renamed to forcing_diag which could impact users
 - two new namelist were added, debug_model (logical) and debug_model_step (integer) and these trigger the debug_ice diagnostics after the timestep associated with debug_model_step.  
 - there is a new file for testing the wghtfile decomposition option, CICE_data/grid/gx1/cice62_gx1_wghtmask.nc.  This file is on cheyenne and I will move it to a few other testing platforms, but maybe it should be added to the gx1 input files.  This file represents the number of months that ice existed at any grid point during the 2005-2009 gx1 production test run and is used to allocate blocks to procs at gx1 with the wghtfile test decomposition.  The maximum value is 120 and these values are read and normalized during the decomposition step in CICE.

These changes address many issues in #437 as well as lines 7-17 and maybe 81-82 in https://docs.google.com/spreadsheets/d/12tIfm5OzvzH_LF_ie9WbOn_w5aY1hv4SrqTQQ8bqM-s.  Code coverage should go up after this update and that will be evaluated separately.

Otherwise, a summary of all changes follows
    
- Add alt06 test, ncat=7, kcatbound=3, nslyr=3
- Add bigdiag test with lots of diagnostic output, debug_model=true, debug_blocks=true, etc
- Add spiralcenter decomp test
- Add gx1 weightfile decomp test
- Add gbox180 with spacecurve decomp test to trigger Peano and Cinco decompositions
- Update sectrobin decomp test to check land block elimination
- Rename boxdyn to boxnodyn
- Update calchk to check 100,000 years instead of 1000 years (which was set for quick debugging, but was not supposed to be checked in).
- Modify boxslotcyl, turn off tr_lvl and tr_pond_lvl, changes answers
- Clarify boxadv, boxnodyn, boxrestore, explicitly set shortwave=ccsm3, albedo_type=constant
- Cleanup spacecurve implementation including debug flagging, recode print to write, module private, add printcurve for gridsize of 30, remove qsort and partition subroutines.
- Deprecate IsLoadBalanced in ice_spacecurve
- Remove CICE_RunMod.F90_debug and migrate high level debug checks to CICE_RunMod.F90 with new debug_model namelist.  Also add debug_model_step to specify at what timestep to start writing debug_model output.  debug_model writes data for the first point specified by lonpnt, latpnt.
- Move debug_ice routine into ice_diagnostics.F90 and out of CICE.F90
- Remove writeout_finished_file from CICE_Finalize in several drivers
- Deprecate print_points_state via an ifdef
- Rename l_stop in ice_transport_driver.F90 to ckflag.  This looked like l_stop logic because of the naming but it isn't.
- Fix issue of size of fcontopn_d and fsurfn_d when ncat greater than 6 in init_coupler_flux.  Would go out of bounds, now initializes fcontopn and fsurfn to 6th value for any category greater than 6.
- Rename dbug variable name in ice_forcing.F90 to forcing_diag and change namelist variable name.
- Update ice_distributionGet output and implementation to support more general arguments
- Add ice_distributionGet test in init_domain_distribution when debug_blocks is true.
- Fix bug in create_distrb_rake for 1d rake of allocation size.  Add a test that triggers the 1d rake.
- Update debug flag in ice_distribution.F90 leverage debug_blocks.
- Update complog part of baseline script to report missing data as MISS instead of FAIL
- Update documentation.
